pkg/service/service.go

```diff
@@ -1,13 +1,125 @@
 package service
 
 import (
+	"context"
+
 	runapi "github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/run"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/config"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/run/v1"
 )
 
-// Client is the connection to update and obtain data about a single service.
+// Client is the connection to update and obtain data about a single service. It
+// contains a client to the Cloud Run API, information about the service and the
+// service object.
 type Client struct {
 	RunClient   runapi.Client
+	Service     *run.Service
 	Project     string
 	ServiceName string
 	Region      string
 }
+
+// NewClient initializes a service client.
+func NewClient(runclient runapi.Client, svc *run.Service, project, serviceName, region string) *Client {
+	return &Client{
+		RunClient:   runclient,
+		Service:     svc,
+		Project:     project,
+		ServiceName: serviceName,
+		Region:      region,
+	}
+}
+
+// Filter retrieves all the services that match the filter in the configuration.
+func Filter(logger *logrus.Logger, cfg *config.Config) []*Client {
+	clientsCh := make(chan []*Client)
```

Comment:
this code is hard to follow wit channels, I'm not sure they are required here. what are you trying to do? 

if all you're trying to do is get services in parallel, you should use WaitGroup + Goroutines; not channel.

Reply:
The major reason this code looks confusing is that I'm using channels to retrieve the service objects I am getting from the API. If I run this in goroutines, what's the best way to store the returned values from the API using a WaitGroup? Using global state? Creating a struct with a `[]run.Service/[]service.Client` field + sync.WaitGroup field? Passing the array storing the values and using mutex?